### PR TITLE
[Key Vault Keys] Added support for OKP keys for sign/verify cryptographic operations on the service side.

### PR DESCRIPTION
**Pull request by vcolin7**

### CODE REVIEW

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/SignatureHashResolver.java [hunk 1]
```diff
@@ -16,14 +16,14 @@ class SignatureHashResolver {
...
         DEFAULT.put(SignatureAlgorithm.ES384, HashAlgorithm.SHA_384);
-
         DEFAULT.put(SignatureAlgorithm.ES512, HashAlgorithm.SHA_512);
         DEFAULT.put(SignatureAlgorithm.RS256, HashAlgorithm.SHA_256);
         DEFAULT.put(SignatureAlgorithm.RS384, HashAlgorithm.SHA_384);
         DEFAULT.put(SignatureAlgorithm.RS512, HashAlgorithm.SHA_512);
         DEFAULT.put(SignatureAlgorithm.PS256, HashAlgorithm.SHA_256);
         DEFAULT.put(SignatureAlgorithm.PS512, HashAlgorithm.SHA_512);
         DEFAULT.put(SignatureAlgorithm.PS384, HashAlgorithm.SHA_384);
+        DEFAULT.put(SignatureAlgorithm.EDDSA, HashAlgorithm.SHA_512);
```
Comment (heaths) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/SignatureHashResolver.java:26]: This seems correct, but we should verify with @daviddesberg, Vick, or anyone from the MHSM team.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeySignRequest.java [hunk 2]
```diff
@@ -17,7 +17,7 @@ public final class KeySignRequest {
      * The signing/verification algorithm. For more information on possible
      * algorithm types, see SignatureAlgorithm. Possible values
      * include: 'PS256', 'PS384', 'PS512', 'RS256', 'RS384', 'RS512', 'RSNULL',
-     * 'ES256', 'ES384', 'ES512', 'ES256K'.
+     * 'ES256', 'ES384', 'ES512', 'ES256K', 'Ed25519'.
```
Comment (heaths) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeySignRequest.java:20]: Shouldn't this be `EdDSA`?

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeyVerifyRequest.java [hunk 3]
```diff
@@ -17,7 +17,7 @@ public final class KeyVerifyRequest {
      * The signing/verification algorithm. For more information on possible
      * algorithm types, see SignatureAlgorithm. Possible values
      * include: 'PS256', 'PS384', 'PS512', 'RS256', 'RS384', 'RS512', 'RSNULL',
-     * 'ES256', 'ES384', 'ES512', 'ES256K'.
+     * 'ES256', 'ES384', 'ES512', 'ES256K', 'Ed25519'.
```
Comment (heaths) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeyVerifyRequest.java:20]: Shouldn't this be `EdDSA`?

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/SignatureHashResolver.java (hunk 1 above)
Comment (heaths) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/SignatureHashResolver.java:26]: Seems this needs to be SHA256. See email conversation for more details.

Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/SignatureHashResolver.java:26]: Will change this to `SHA_256` as that is the only algorithm currently supported by the service.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeySignRequest.java (hunk 2 above)
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeySignRequest.java:20]: No, the services specifies `Ed25519` as the curve name [here](https://github.com/Azure/azure-rest-api-specs/blob/21826324d0c76ef0309fe7761eaa34d4da174359/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/keys.json#L1386).

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeyVerifyRequest.java (hunk 3 above)
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeyVerifyRequest.java:20]: No, the services specifies `Ed25519` as the curve name [here](https://github.com/Azure/azure-rest-api-specs/blob/21826324d0c76ef0309fe7761eaa34d4da174359/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/keys.json#L1386).

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/SignatureHashResolver.java [hunk 4]
```diff
@@ -16,14 +16,15 @@ class SignatureHashResolver {
...
         DEFAULT.put(SignatureAlgorithm.ES384, HashAlgorithm.SHA_384);
-
         DEFAULT.put(SignatureAlgorithm.ES512, HashAlgorithm.SHA_512);
         DEFAULT.put(SignatureAlgorithm.RS256, HashAlgorithm.SHA_256);
         DEFAULT.put(SignatureAlgorithm.RS384, HashAlgorithm.SHA_384);
         DEFAULT.put(SignatureAlgorithm.RS512, HashAlgorithm.SHA_512);
         DEFAULT.put(SignatureAlgorithm.PS256, HashAlgorithm.SHA_256);
         DEFAULT.put(SignatureAlgorithm.PS512, HashAlgorithm.SHA_512);
         DEFAULT.put(SignatureAlgorithm.PS384, HashAlgorithm.SHA_384);
+        // The specification says we should use SHA_512 but the service only supports SHA-256 for now for EdDSA.
```
Comment (pallavit) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/SignatureHashResolver.java:26]: Do you want to add a TODO: to add a follow-up when the service does support SHA256?

**Comment by pallavit** [comment]
>      * Elliptic curve name. For valid values, see KeyCurveName. Possible

Should you not add Ed25519 here?

---
Refers to: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/JsonWebKey.java:142 in e244edc. [](commit_id = e244edcdf81954ba5bbbe5b0f1162faaad08fbe6, deletion_comment = False)

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeyVerifyRequest.java (hunk 3 above)
Comment (pallavit) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeyVerifyRequest.java:20]: That is for curve this one signature algorithm - [https://github.com/Azure/azure-rest-api-specs/blob/21826324d0c76ef0309fe7761eaa34d4da174359/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/keys.json#L1901](https://github.com/Azure/azure-rest-api-specs/blob/21826324d0c76ef0309fe7761eaa34d4da174359/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/keys.json#L1901)

Comment (pallavit) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeyVerifyRequest.java:20]: That one is for curve Eddsa would be for signature algorithm - [https://github.com/Azure/azure-rest-api-specs/blob/21826324d0c76ef0309fe7761eaa34d4da174359/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/keys.json#L1983](https://github.com/Azure/azure-rest-api-specs/blob/21826324d0c76ef0309fe7761eaa34d4da174359/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/keys.json#L1983)

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeySignRequest.java (hunk 2 above)
Comment (pallavit) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeySignRequest.java:20]: That one is for the curve - signature algorithm values are EdDSA only. [https://github.com/Azure/azure-rest-api-specs/blob/21826324d0c76ef0309fe7761eaa34d4da174359/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/keys.json#L1901](https://github.com/Azure/azure-rest-api-specs/blob/21826324d0c76ef0309fe7761eaa34d4da174359/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/keys.json#L1901)

**Comment by pallavit** [comment]
Added a few comments but otherwise looks good to me.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeyVerifyRequest.java (hunk 3 above)
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeyVerifyRequest.java:20]: You're both right, my bad! :P

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeySignRequest.java (hunk 2 above)
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/implementation/KeySignRequest.java:20]: You're both right, my bad! :P

**Review by pallavit — APPROVED** [review]

**Pull request merged**